src/CPUActivityAgent.cpp

```diff
@@ -0,0 +1,564 @@
+/*
+ * Copyright (c) 2015 - 2022, Intel Corporation
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include "CPUActivityAgent.hpp"
+
+#include <cmath>
+#include <cassert>
+#include <algorithm>
+#include <iostream>
+#include <string>
+
+#include "geopm/Agg.hpp"
+#include "geopm/Exception.hpp"
+#include "geopm/Helper.hpp"
+#include "geopm/PlatformIO.hpp"
+#include "geopm/PlatformTopo.hpp"
+#include "geopm/PluginFactory.hpp"
+#include "geopm_debug.hpp"
+
+#include "PlatformIOProf.hpp"
+
+namespace geopm
+{
+    CPUActivityAgent::CPUActivityAgent()
+        : CPUActivityAgent(platform_io(), platform_topo())
+    {
+
+    }
+
+    CPUActivityAgent::CPUActivityAgent(PlatformIO &plat_io, const PlatformTopo &topo)
+        : m_platform_io(plat_io)
+        , m_platform_topo(topo)
+        , m_last_wait{{0, 0}}
+        , M_WAIT_SEC(0.010) // 10ms wait default
+        , M_POLICY_PHI_DEFAULT(0.5)
+        , M_NUM_PACKAGE(m_platform_topo.num_domain(GEOPM_DOMAIN_PACKAGE))
+        , M_NUM_CORE(m_platform_topo.num_domain(GEOPM_DOMAIN_CORE))
+        , m_do_write_batch(false)
+        , m_do_send_policy(true)
+        , m_core_frequency_requests(0)
+        , m_uncore_frequency_requests(0)
+        , m_core_frequency_clipped(0)
+        , m_uncore_frequency_clipped(0)
+        , m_resolved_f_uncore_efficient(0)
+        , m_resolved_f_uncore_max(0)
+        , m_resolved_f_core_efficient(0)
+        , m_resolved_f_core_max(0)
+    {
+        geopm_time(&m_last_wait);
+    }
+
+    // Push signals and controls for future batch read/write
+    void CPUActivityAgent::init(int level, const std::vector<int> &fan_in, bool is_level_root)
+    {
+        // These are not currently guaranteed to be the system uncore min and max,
+        // just what the user/admin has previously set.
+        m_freq_uncore_min = m_platform_io.read_signal("CPU_UNCORE_FREQUENCY_MIN_CONTROL", GEOPM_DOMAIN_BOARD, 0);
+        m_freq_uncore_max = m_platform_io.read_signal("CPU_UNCORE_FREQUENCY_MAX_CONTROL", GEOPM_DOMAIN_BOARD, 0);
+        m_freq_core_min = m_platform_io.read_signal("CPU_FREQUENCY_MIN_AVAIL", GEOPM_DOMAIN_BOARD, 0);
+        m_freq_core_max = m_platform_io.read_signal("CPU_FREQUENCY_MAX_AVAIL", GEOPM_DOMAIN_BOARD, 0);
+
+        if (level == 0) {
+            init_platform_io();
+        }
+    }
+
+    void CPUActivityAgent::init_platform_io(void)
+    {
+        for (int domain_idx = 0; domain_idx < M_NUM_CORE; ++domain_idx) {
+            m_core_scal.push_back({m_platform_io.push_signal("MSR::CPU_SCALABILITY_RATIO",
+                                                             GEOPM_DOMAIN_CORE,
+                                                             domain_idx), NAN});
+            m_core_freq_control.push_back({m_platform_io.push_control("CPU_FREQUENCY_MAX_CONTROL",
+                                                                      GEOPM_DOMAIN_CORE,
+                                                                      domain_idx), -1});
```

Comment:
Why -1 instead of NaN?

Reply:
In ``CPUActivityAgent::adjust_platform()``, when it determines what the frequency should be (``core_request``), it needs to compare it against the last setting to decide whether the setting actually needs to be updated. Using -1 allows that check to work properly the first time a frequency decision is made.